File: ios/NfcKitModule.swift
import ExpoModulesCore
import UIKit
@preconcurrency import CoreNFC

/// Bumped together with `CONTRACT_VERSION` in `src/native/contract.ts`.
private let contractVersion = 1

/// Mirrors `NativeSessionOptions`. The Android fields are accepted and ignored.
internal struct SessionOptions: Record {
  @Field var techs: [String] = []
  @Field var iosPollingOptions: [String]?
  @Field var iosAlertMessage: String?
  @Field var iosInvalidateAfterFirstRead: Bool = false
  @Field var iosSelectIdentifiers: [String]?
  @Field var iosFelicaSystemCodes: [String]?
  @Field var androidSkipNdefCheck: Bool = false
  @Field var androidNoPlatformSounds: Bool = false
  @Field var androidPresenceCheckDelayMs: Int?
}

public final class NfcKitModule: Module {
  /// Created lazily so the event sink can capture `self` safely.
  private var coordinatorStorage: NfcSessionCoordinator?

  private func coordinator() -> NfcSessionCoordinator {
    if let coordinatorStorage {
      return coordinatorStorage
    }
    let created = NfcSessionCoordinator { [weak self] name, body in
      self?.sendEvent(name, body)
    }
    coordinatorStorage = created
    return created
  }

  public func definition() -> ModuleDefinition {
    Name("NfcKit")

    Constant("contractVersion") { contractVersion }

    Constant("capabilities") {
      [
        "platform": "ios",
        "osVersion": UIDevice.current.systemVersion,
        // What CoreNFC can reach. MIFARE Classic is absent because Crypto-1 is
        // not implemented on iOS at any OS version, and raw NfcA/NfcB are absent
        // because CoreNFC offers no framing at that level. Saying so here is what
        // makes tag.is('mifareClassic') honestly false rather than a method that
        // exists and then fails.
        "techs": ["ndef", "isoDep", "iso15693", "felica", "mifareUltralight"],
        // iOS reports no tag-removal callback at all.
        "nativeTagLost": false,
        // iOS 26.4 can narrow AIDs and FeliCa system codes per session. Adopting
        // it needs a build SDK that has NFCTagReaderSession.Configuration, so
        // this stays false until that lands rather than implying narrowing that
        // does not happen.
        "perSessionConfig": false,
        "hce": false,
        "backgroundReading": false
      ] as [String: Any]
    }

    Events("onTagDiscovered", "onTagLost", "onSessionInvalidated", "onAvailabilityChanged")

    /* -- Availability ---------------------------------------------------- */

    AsyncFunction("isSupported") { () -> Bool in
      NFCTagReaderSession.readingAvailable
    }

    AsyncFunction("isEnabled") { () -> Bool in
      // iOS has no user-facing NFC toggle, so "enabled" means the same as
      // "supported". Reporting true unconditionally, as the previous library
      // did, claims a working radio on an iPad.
      NFCTagReaderSession.readingAvailable
    }

    AsyncFunction("openSettings") { () -> Void in
      throw NfcException(
        NfcErrorCode.unsupportedPlatform,
        "iOS has no NFC settings screen and no way to link to one. NFC is always on when the "
          + "device supports it."
      )
    }

    /* -- Session lifecycle ----------------------------------------------- */

    AsyncFunction("startSession") { (sessionId: String, options: SessionOptions) in
      try await self.coordinator().start(
        sessionId: sessionId,
        techs: options.techs,
        alertMessage: options.iosAlertMessage
      )
    }

    AsyncFunction("closeSession") { (sessionId: String) in
      await self.coordinator().close(sessionId: sessionId)
    }

    AsyncFunction("setSessionAlert") { (sessionId: String, message: String) in
      await self.coordinator().setAlert(sessionId: sessionId, message: message)
    }

    /* -- Tag operations --------------------------------------------------- */

    AsyncFunction("releaseTag") { (handleId: String) in
      await self.coordinator().releaseTag(handleId)
    }

    AsyncFunction("readNdef") { (handleId: String) -> Data in
      try await self.coordinator().handle(handleId).readNdef()
    }

    AsyncFunction("writeNdef") { (handleId: String, message: Data) in
      try await self.coordinator().handle(handleId).writeNdef(message)
    }

    AsyncFunction("getNdefStatus") { (handleId: String) -> [String: Any?] in
      try await self.coordinator().handle(handleId).ndefStatusPayload()
    }

    AsyncFunction("makeNdefReadOnly") { (handleId: String) in
      try await self.coordinator().handle(handleId).makeNdefReadOnly()
    }

    AsyncFunction("formatNdef") { (_: String, _: Data) in
      // CoreNFC has no formatting API: a tag arrives either NDEF-capable or not.
      throw NfcException(
        NfcErrorCode.unsupportedPlatform,
        "CoreNFC cannot format a tag for NDEF. Format it on Android, or with a tag writer, and "
          + "iOS will then read and write it."
      )
    }

    AsyncFunction("transceive") { (handleId: String, tech: String, data: Data) -> Data in
      try await self.coordinator().handle(handleId).transceive(tech: tech, data: data)
    }

    AsyncFunction("getMaxTransceiveLength") { (handleId: String, tech: String) -> Int in
      try await self.coordinator().handle(handleId).maxTransceiveLength(tech: tech)
    }

    AsyncFunction("setTechTimeout") { (_: String, _: String, _: Int) in
      throw NfcException(
        NfcErrorCode.unsupportedPlatform,
        "CoreNFC does not expose per-technology timeouts. Guard this call with tag.android."
      )
    }

    AsyncFunction("getTechTimeout") { (_: String, _: String) -> Int in
      throw NfcException(
        NfcErrorCode.unsupportedPlatform,
        "CoreNFC does not expose per-technology timeouts. Guard this call with tag.android."
      )
    }

    AsyncFunction("takeLaunchTag") { () -> [String: Any?]? in
      // Background NDEF hand-off lands in a later milestone. Reporting nil is
      // accurate today; inventing a tag would not be.
      nil
    }

    /* -- Lifecycle -------------------------------------------------------- */

    OnDestroy {
      // A session outliving the module would leave the system sheet up with
      // nothing behind it.
      if let coordinator = self.coordinatorStorage {
        Task { await coordinator.closeAny() }
      }
    }
  }
}

File: android/src/main/java/com/nfckit/NfcKitModule.kt
package com.nfckit

import android.app.Activity
import android.content.Context
import android.content.Intent
import android.nfc.NfcAdapter
import android.os.Build
import android.provider.Settings
import com.nfckit.reader.ReaderModeController
import com.nfckit.reader.ReaderOptions
import com.nfckit.reader.TagHandle
import com.nfckit.reader.TagIo
import com.nfckit.tech.TechRegistry
import expo.modules.kotlin.functions.Coroutine
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import expo.modules.kotlin.records.Field
import expo.modules.kotlin.records.Record
import kotlinx.coroutines.withContext
import java.util.concurrent.ConcurrentHashMap
import java.util.concurrent.atomic.AtomicInteger

/** Bumped together with `CONTRACT_VERSION` in `src/native/contract.ts`. */
private const val CONTRACT_VERSION = 1

private const val EVENT_TAG_DISCOVERED = "onTagDiscovered"
private const val EVENT_TAG_LOST = "onTagLost"
private const val EVENT_SESSION_INVALIDATED = "onSessionInvalidated"
private const val EVENT_AVAILABILITY_CHANGED = "onAvailabilityChanged"

/** Mirrors `NativeSessionOptions`. The iOS fields are accepted and ignored here. */
class SessionOptions : Record {
  @Field val techs: List<String> = emptyList()
  @Field val iosPollingOptions: List<String>? = null
  @Field val iosAlertMessage: String? = null
  @Field val iosInvalidateAfterFirstRead: Boolean = false
  @Field val iosSelectIdentifiers: List<String>? = null
  @Field val iosFelicaSystemCodes: List<String>? = null
  @Field val androidSkipNdefCheck: Boolean = false
  @Field val androidNoPlatformSounds: Boolean = false
  @Field val androidPresenceCheckDelayMs: Int? = null
}

/** One open reading session and the tags it has handed out. */
private class ReaderSession(val id: String, val techs: List<String>) {
  val handles = ConcurrentHashMap<String, TagHandle>()
  private val counter = AtomicInteger(0)

  fun nextHandleId(): String = "$id-t${counter.incrementAndGet()}"

  fun closeAll() {
    handles.values.forEach { it.close() }
    handles.clear()
  }
}

class NfcKitModule : Module() {

  private var session: ReaderSession? = null

  private val readerMode = ReaderModeController { nfcAdapter() }

  /* ---------------------------------------------------------------------- */
  /* Context access                                                         */
  /* ---------------------------------------------------------------------- */

  /**
   * The current activity, fetched every time and never cached.
   *
   * Caching a `Context` or an `Activity` is exactly the bug that took the library
   * this replaces two releases to unwind for React Native 0.80 and 0.81: the
   * cached reference went stale across activity recreation, so `isSupported` and
   * `isEnabled` reported the wrong thing until the app restarted.
   *
   * "No foreground activity" is a first-class error rather than a string returned
   * from five different methods, because reader mode genuinely requires one.
   */
  private fun requireActivity(): Activity =
    appContext.activityProvider?.currentActivity
      ?: throw NfcException(
        NfcErrorCode.NO_ACTIVITY,
        "NFC reader mode needs a foreground activity and there is none. This usually means the " +
          "app is in the background, or the scan was started before the activity was attached.",
      )

  private fun currentActivityOrNull(): Activity? = appContext.activityProvider?.currentActivity

  private fun nfcAdapter(): NfcAdapter? {
    val context: Context = appContext.reactContext ?: return null
    return NfcAdapter.getDefaultAdapter(context)
  }

  /* ---------------------------------------------------------------------- */
  /* Definition                                                             */
  /* ---------------------------------------------------------------------- */

  override fun definition() = ModuleDefinition {
    Name("NfcKit")

    Constant("contractVersion") { CONTRACT_VERSION }

    Constant("capabilities") {
      mapOf(
        "platform" to "android",
        "osVersion" to Build.VERSION.RELEASE,
        // Every technology this build can drive. Which of them a given tag
        // actually supports is a per-tag question -- MIFARE Classic is
        // chipset-dependent -- so `tag.techs` is the authoritative answer.
        "techs" to TechRegistry.allTechNames,
        // Tag removal is not reported yet on Android. Android 17 (API 37) adds
        // ReaderCallback.onTagLost; until this module compiles against that SDK,
        // saying false is the truth rather than shipping an unreliable poller.
        "nativeTagLost" to false,
        "perSessionConfig" to false,
        "hce" to false,
        "backgroundReading" to false,
      )
    }

    Events(
      EVENT_TAG_DISCOVERED,
      EVENT_TAG_LOST,
      EVENT_SESSION_INVALIDATED,
      EVENT_AVAILABILITY_CHANGED,
    )

    /* -- Availability ---------------------------------------------------- */

    AsyncFunction("isSupported") { nfcAdapter() != null }

    AsyncFunction("isEnabled") { nfcAdapter()?.isEnabled == true }

    AsyncFunction("openSettings") {
      val activity = requireActivity()
      activity.startActivity(Intent(Settings.ACTION_NFC_SETTINGS))
    }

    /* -- Session lifecycle ----------------------------------------------- */

    AsyncFunction("startSession") { sessionId: String, options: SessionOptions ->
      val adapter = readerMode.requireAdapter()
      if (!adapter.isEnabled) {
        throw NfcException(
          NfcErrorCode.NFC_DISABLED,
          "NFC is switched off. Ask the user to enable it, or call openSettings() to take them there.",
        )
      }
      if (session != null) {
        throw NfcException(
          NfcErrorCode.SYSTEM_BUSY,
          "An NFC session is already open. Close it before starting another.",
        )
      }

      val activity = requireActivity()
      val active = ReaderSession(sessionId, options.techs)
      session = active

      try {
        readerMode.want(
          activity = activity,
          techs = options.techs,
          options = ReaderOptions(
            skipNdefCheck = options.androidSkipNdefCheck,
            noPlatformSounds = options.androidNoPlatformSounds,
            presenceCheckDelayMs = options.androidPresenceCheckDelayMs,
          ),
          callback = { tag -> onTagDiscovered(tag) },
        )
      } catch (error: Throwable) {
        // Never leave a half-started session behind: the next startSession would
        // report systemBusy for a session that never actually began.
        session = null
        readerMode.unwant(currentActivityOrNull())
        throw error
      }
    }

    AsyncFunction("closeSession") { sessionId: String ->
      val active = session
      if (active != null && active.id == sessionId) {
        session = null
        readerMode.unwant(currentActivityOrNull())
        active.closeAll()
      }
      // Closing an already-closed session is not an error. JavaScript closes in a
      // finally block, which can legitimately run after the platform already
      // ended the session.
    }

    AsyncFunction("setSessionAlert") { _: String, _: String ->
      // iOS shows a system sheet whose text can be updated mid-scan. Android has
      // no equivalent, and the app owns its own UI, so this is a no-op rather
      // than an error: cross-platform code should not have to branch for it.
    }

    /* -- Tag operations --------------------------------------------------- */

    AsyncFunction("releaseTag") Coroutine { handleId: String ->
      withContext(TagIo.dispatcher) {
        session?.handles?.remove(handleId)?.close()
      }
    }

    AsyncFunction("readNdef") Coroutine { handleId: String ->
      onTagIo(handleId) { it.readNdef() }
    }

    AsyncFunction("writeNdef") Coroutine { handleId: String, message: ByteArray ->
      onTagIo(handleId) { it.writeNdef(message) }
    }

    AsyncFunction("getNdefStatus") Coroutine { handleId: String ->
      onTagIo(handleId) { handle ->
        val status = handle.ndefStatus()
        mapOf(
          "writable" to status.writable,
          "capacity" to status.capacity,
          "canMakeReadOnly" to status.canMakeReadOnly,
          "typeName" to status.typeName,
        )
      }
    }

    AsyncFunction("makeNdefReadOnly") Coroutine { handleId: String ->
      onTagIo(handleId) { it.makeNdefReadOnly() }
    }

    AsyncFunction("formatNdef") Coroutine { handleId: String, message: ByteArray ->
      onTagIo(handleId) { it.formatNdef(message) }
    }

    AsyncFunction("transceive") Coroutine { handleId: String, tech: String, data: ByteArray ->
      onTagIo(handleId) { it.transceive(tech, data) }
    }

    AsyncFunction("getMaxTransceiveLength") Coroutine { handleId: String, tech: String ->
      onTagIo(handleId) { it.maxTransceiveLength(tech) }
    }

    AsyncFunction("setTechTimeout") Coroutine { handleId: String, tech: String, timeoutMs: Int ->
      onTagIo(handleId) { it.setTechTimeout(tech, timeoutMs) }
    }

    AsyncFunction("getTechTimeout") Coroutine { handleId: String, tech: String ->
      onTagIo(handleId) { it.getTechTimeout(tech) }
    }

    AsyncFunction("takeLaunchTag") {
      // Background tag delivery lands in a later milestone. Reporting null is
      // accurate today; inventing a tag would not be. The explicit type keeps
      // the return from inferring as Nothing?.
      null as Map<String, Any?>?
    }

    /* -- Lifecycle -------------------------------------------------------- */

    OnActivityEntersForeground {
      // Reader mode belongs to the foreground activity, so it has to be
      // re-established after the app comes back. The session survives; only the
      // radio was released.
      currentActivityOrNull()?.let { readerMode.resumeIfWanted(it) }
    }

    OnActivityEntersBackground {
      readerMode.suspend(currentActivityOrNull())
    }

    OnActivityDestroys {
      teardown()
    }

    OnDestroy {
      teardown()
    }
  }

  /* ---------------------------------------------------------------------- */
  /* Helpers                                                                */
  /* ---------------------------------------------------------------------- */

  /** Runs a tag operation on the dedicated I/O thread. */
  private suspend fun <T> onTagIo(handleId: String, block: (TagHandle) -> T): T =
    withContext(TagIo.dispatcher) {
      val handle = session?.handles?.get(handleId)
        ?: throw NfcException(
          NfcErrorCode.SESSION_CLOSED,
          "This tag is no longer available: its session has been closed.",
        )
      block(handle)
    }

  /**
   * Called on the NFC service's own thread when a tag enters the field.
   *
   * Everything here is cheap and non-blocking -- reading a tech list touches no
   * radio -- so it is safe to do inline. The first actual exchange happens later,
   * on the tag-I/O thread, when JavaScript asks for something.
   */
  private fun onTagDiscovered(tag: android.nfc.Tag) {
    val active = session ?: return
    val handle = TagHandle(active.nextHandleId(), tag)
    active.handles[handle.id] = handle

    sendEvent(
      EVENT_TAG_DISCOVERED,
      mapOf(
        "sessionId" to active.id,
        "tag" to mapOf(
          "handleId" to handle.id,
          "idHex" to handle.idHex,
          "techs" to handle.techs,
          "android" to mapOf(
            "techList" to handle.androidTechList,
            "maxTransceiveLength" to null,
            "hiLayerResponseHex" to null,
            "historicalBytesHex" to null,
          ),
          "ios" to null,
        ),
      ),
    )
  }

  private fun teardown() {
    val active = session
    session = null
    readerMode.unwant(currentActivityOrNull())
    active?.closeAll()
  }
}
